docs/man/chasquid.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "chasquid 1"
.TH chasquid 1 2023-10-03 "" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
chasquid \- SMTP (email) server
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBchasquid\fR [\fIoptions\fR...]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
chasquid is an SMTP (email) server with a focus on simplicity, security, and
ease of operation.
.PP
It\*(Aqs written in Go, and distributed under the Apache license 2.0.
.SH OPTIONS
.IX Header "OPTIONS"
.IP "\fB\-config_dir\fR \fIdir\fR" 8
.IX Item "-config_dir dir"
configuration directory (default \fI/etc/chasquid\fR)
.IP "\fB\-config_overrides\fR \fIconfig\fR" 8
.IX Item "-config_overrides config"
configuration values (in text protobuf format) to override the on\-disk
configuration with. This should only be needed in very specific cases for
deployments where editing the configuration file is not feasible.
.IP \fB\-alsologtostderr\fR 8
.IX Item "-alsologtostderr"
also log to stderr, in addition to the file
.IP "\fB\-logfile\fR \fIfile\fR" 8
.IX Item "-logfile file"
file to log to (enables logtime)
.IP \fB\-logtime\fR 8
.IX Item "-logtime"
include the time when writing the log to stderr
.IP "\fB\-logtosyslog\fR \fItag\fR" 8
.IX Item "-logtosyslog tag"
log to syslog, with the given tag
.IP "\fB\-v\fR \fIlevel\fR" 8
.IX Item "-v level"
verbosity level (1 = debug)
.IP \fB\-version\fR 8
.IX Item "-version"
show version and exit
.SH FILES
.IX Header "FILES"
The daemon\*(Aqs configuration is by default in \fI/etc/chasquid/\fR, and can be
changed with the \fI\-config_dir\fR flag.
.PP
Inside that directory, the daemon expects the following structure:
.IP \fIchasquid.conf\fR 8
.IX Item "chasquid.conf"
Main config file, see \fBchasquid.conf\fR\|(5).
.IP \fIdomains/\fR 8
.IX Item "domains/"
Per\-domain configuration.
.IP \fIdomains/example.com/\fR 8
.IX Item "domains/example.com/"
Domain\-specific configuration. Can be empty.
.IP \fIdomains/example.com/users\fR 8
.IX Item "domains/example.com/users"
User and password database for this domain.
.IP \fIdomains/example.com/aliases\fR 8
.IX Item "domains/example.com/aliases"
Aliases for the domain.
.IP \fIcerts/\fR 8
.IX Item "certs/"
Certificates to use, one directory per pair.
.IP \fIcerts/mx.example.com/\fR 8
.IX Item "certs/mx.example.com/"
Certificates for this domain.
.IP \fIcerts/mx.example.com/fullchain.pem\fR 8
.IX Item "certs/mx.example.com/fullchain.pem"
Certificate (full chain).
.IP \fIcerts/mx.example.com/privkey.pem\fR 8
.IX Item "certs/mx.example.com/privkey.pem"
Private key.
.PP
Note the \fIcerts/\fR directory layout matches the one from certbot (client for
Let\*(Aqs Encrypt CA), so you can just symlink \fIcerts/\fR to
\&\fI/etc/letsencrypt/live\fR.
.PP
Make sure the user you use to run chasquid under ("mail" in the example
config) can access the certificates and private keys.
.SH CONTACT
.IX Header "CONTACT"
Main website <https://blitiri.com.ar/p/chasquid>.
.PP
If you have any questions, comments or patches please send them to the mailing
list, \f(CW\*(C`chasquid@googlegroups.com\*(C'\fR.  To subscribe, send an email to
\&\f(CW\*(C`chasquid+subscribe@googlegroups.com\*(C'\fR.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBchasquid\-util\fR\|(1), \fBchasquid.conf\fR\|(5)


chasquid.go
// chasquid is an SMTP (email) server, with a focus on simplicity, security,
// and ease of operation.
//
// See https://blitiri.com.ar/p/chasquid for more details.
package main

import (
	"context"
	"errors"
	"flag"
	"fmt"
	"net"
	"os"
	"os/signal"
	"path"
	"path/filepath"
	"strings"
	"syscall"
	"time"

	"blitiri.com.ar/go/chasquid/internal/config"
	"blitiri.com.ar/go/chasquid/internal/courier"
	"blitiri.com.ar/go/chasquid/internal/domaininfo"
	"blitiri.com.ar/go/chasquid/internal/dovecot"
	"blitiri.com.ar/go/chasquid/internal/localrpc"
	"blitiri.com.ar/go/chasquid/internal/maillog"
	"blitiri.com.ar/go/chasquid/internal/normalize"
	"blitiri.com.ar/go/chasquid/internal/smtpsrv"
	"blitiri.com.ar/go/chasquid/internal/sts"
	"blitiri.com.ar/go/log"
	"blitiri.com.ar/go/systemd"
)

// Command-line flags.
var (
	configDir = flag.String("config_dir", "/etc/chasquid",
		"configuration directory")
	configOverrides = flag.String("config_overrides", "",
		"override configuration values (in text protobuf format)")
	showVer = flag.Bool("version", false, "show version and exit")
)

func main() {
	flag.Parse()
	log.Init()

	parseVersionInfo()
	if *showVer {
		fmt.Printf("chasquid %s (source date: %s)\n", version, sourceDate)
		return
	}

	log.Infof("chasquid starting (version %s)", version)

	conf, err := config.Load(*configDir+"/chasquid.conf", *configOverrides)
	if err != nil {
		log.Fatalf("Error loading config: %v", err)
	}
	config.LogConfig(conf)

	// Change to the config dir.
	// This allow us to use relative paths for configuration directories.
	// It also can be useful in unusual environments and for testing purposes,
	// where paths inside the configuration itself could be relative, and this
	// fixes the point of reference.
	err = os.Chdir(*configDir)
	if err != nil {
		log.Fatalf("Error changing to config dir %q: %v", *configDir, err)
	}

	initMailLog(conf.MailLogPath)

	if conf.MonitoringAddress != "" {
		go launchMonitoringServer(conf)
	}

	s := smtpsrv.NewServer()
	s.Hostname = conf.Hostname
	s.MaxDataSize = conf.MaxDataSizeMb * 1024 * 1024
	s.HookPath = "hooks/"
	s.HAProxyEnabled = conf.HaproxyIncoming

	s.SetAliasesConfig(*conf.SuffixSeparators, *conf.DropCharacters)

	if conf.DovecotAuth {
		loadDovecot(s, conf.DovecotUserdbPath, conf.DovecotClientPath)
	}

	// Load certificates from "certs/<directory>/{fullchain,privkey}.pem".
	// The structure matches letsencrypt's, to make it easier for that case.
	log.Infof("Loading certificates:")
	for _, info := range mustReadDir("certs/") {
		if info.Type().IsRegular() {
			// Ignore regular files, we only care about directories.
			continue
		}

		name := info.Name()
		dir := filepath.Join("certs/", name)
		loadCert(name, dir, s)
	}

	// Load domains from "domains/".
	log.Infof("Loading domains:")
	for _, info := range mustReadDir("domains/") {
		domain, err := normalize.Domain(info.Name())
		if err != nil {
			log.Fatalf("Invalid name %+q: %v", info.Name(), err)
		}
		if info.Type().IsRegular() {
			// Ignore regular files, we only care about directories.
			continue
		}
		dir := filepath.Join("domains", info.Name())
		loadDomain(domain, dir, s)
	}

	// Always include localhost as local domain.
	// This can prevent potential trouble if we were to accidentally treat it
	// as a remote domain (for loops, alias resolutions, etc.).
	s.AddDomain("localhost")

	dinfo, err := domaininfo.New(conf.DataDir + "/domaininfo")
	if err != nil {
		log.Fatalf("Error opening domain info database: %v", err)
	}
	s.SetDomainInfo(dinfo)

	stsCache, err := sts.NewCache(conf.DataDir + "/sts-cache")
	if err != nil {
		log.Fatalf("Failed to initialize STS cache: %v", err)
	}
	go stsCache.PeriodicallyRefresh(context.Background())

	localC := &courier.MDA{
		Binary:  conf.MailDeliveryAgentBin,
		Args:    conf.MailDeliveryAgentArgs,
		Timeout: 30 * time.Second,
	}
	remoteC := &courier.SMTP{
		HelloDomain: conf.Hostname,
		Dinfo:       dinfo,
		STSCache:    stsCache,
	}
	s.InitQueue(conf.DataDir+"/queue", localC, remoteC)
	s.SetQueueLimits(conf.MaxQueueItems, conf.GiveUpSendAfterDuration())

	// Load the addresses and listeners.
	systemdLs, err := systemd.Listeners()
	if err != nil {
		log.Fatalf("Error getting systemd listeners: %v", err)
	}

	naddr := loadAddresses(s, conf.SmtpAddress,
		systemdLs["smtp"], smtpsrv.ModeSMTP)
	naddr += loadAddresses(s, conf.SubmissionAddress,
		systemdLs["submission"], smtpsrv.ModeSubmission)
	naddr += loadAddresses(s, conf.SubmissionOverTlsAddress,
		systemdLs["submission_tls"], smtpsrv.ModeSubmissionTLS)

	if naddr == 0 {
		log.Fatalf("No address to listen on")
	}

	go localrpc.DefaultServer.ListenAndServe(conf.DataDir + "/localrpc-v1")

	go signalHandler(dinfo, s)

	s.ListenAndServe()
}

func loadAddresses(srv *smtpsrv.Server, addrs []string, ls []net.Listener, mode smtpsrv.SocketMode) int {
	naddr := 0
	for _, addr := range addrs {
		if addr == "" {
			// An empty address is invalid, to prevent accidental
			// misconfiguration.
			log.Errorf("Invalid empty listening address for %v", mode)
			log.Fatalf("If you want to disable %v, remove it from the config", mode)
		} else if addr == "systemd" {
			// The "systemd" address indicates we get listeners via systemd.
			srv.AddListeners(ls, mode)
			naddr += len(ls)
		} else {
			srv.AddAddr(addr, mode)
			naddr++
		}
	}

	if naddr == 0 {
		log.Errorf("Warning: No %v addresses/listeners", mode)
		log.Errorf("If using systemd, check that you named the sockets")
	}
	return naddr
}

func initMailLog(path string) {
	var err error

	switch path {
	case "<syslog>":
		maillog.Default, err = maillog.NewSyslog()
	case "<stdout>":
		maillog.Default = maillog.New(os.Stdout)
	case "<stderr>":
		maillog.Default = maillog.New(os.Stderr)
	default:
		_ = os.MkdirAll(filepath.Dir(path), 0775)
		maillog.Default, err = maillog.NewFile(path)
	}

	if err != nil {
		log.Fatalf("Error opening mail log: %v", err)
	}
}

func signalHandler(dinfo *domaininfo.DB, srv *smtpsrv.Server) {
	var err error

	signals := make(chan os.Signal, 1)
	signal.Notify(signals, syscall.SIGHUP, syscall.SIGTERM, syscall.SIGINT)

	for {
		switch sig := <-signals; sig {
		case syscall.SIGHUP:
			log.Infof("Received SIGHUP, reloading")

			// SIGHUP triggers a reopen of the log files. This is used for log
			// rotation.
			err = log.Default.Reopen()
			if err != nil {
				log.Fatalf("Error reopening log: %v", err)
			}

			err = maillog.Default.Reopen()
			if err != nil {
				log.Fatalf("Error reopening maillog: %v", err)
			}

			// We don't want to reload the domain info database periodically,
			// as it can be expensive, and it is not expected that the user
			// changes this behind chasquid's back.
			err = dinfo.Reload()
			if err != nil {
				log.Fatalf("Error reloading domain info: %v", err)
			}

			// Also trigger a server reload.
			srv.Reload()
		case syscall.SIGTERM, syscall.SIGINT:
			log.Infof("Got signal to exit: %v", sig)
			// Ideally, we would shutdown the server gracefully, but for now
			// we just exit. Note that in practice this is not a significant
			// problem, as any in-flight transaction should be retried.
			os.Exit(0)
		default:
			log.Errorf("Unexpected signal %v", sig)
		}
	}
}

// Helper to load a single certificate configuration into the server.
func loadCert(name, dir string, s *smtpsrv.Server) {
	log.Infof("  %s", name)

	// Ignore directories that don't have both keys.
	// We warn about this because it can be hard to debug otherwise.
	certPath := filepath.Join(dir, "fullchain.pem")
	if _, err := os.Stat(certPath); err != nil {
		log.Infof("    skipping: %v", err)
		return
	}

	keyPath, err := findCertKey(dir)
	if err != nil {
		log.Infof("    skipping: %v", err)
		return
	}

	err = s.AddCerts(certPath, keyPath)
	if err != nil {
		log.Fatalf("    %v", err)
	}
}

func findCertKey(dir string) (string, error) {
	keyFiles := []string{
		// Letsencrypt Certbot.
		"privkey.pem",
		// Lego (NixOS ACME module).
		"key.pem",
	}

	for _, name := range keyFiles {
		keyPath := filepath.Join(dir, name)
		_, err := os.Stat(keyPath)
		if err == nil {
			return keyPath, nil
		}
		if !errors.Is(err, os.ErrNotExist) {
			// If the file does not exist, we move on to try the next file.
			// But if it exists but it's inaccessible, return the error so
			// users can debug it more clearly.
			return "", err
		}
	}

	return "", fmt.Errorf(
		"no private key found in %q (tried %v)", dir, keyFiles)
}

// Helper to load a single domain configuration into the server.
func loadDomain(name, dir string, s *smtpsrv.Server) {
	s.AddDomain(name)

	nusers, err := s.AddUserDB(name, dir+"/users")
	if err != nil {
		// If there is an error loading users, fail hard to make sure this is
		// noticed and fixed as soon as it happens.
		log.Fatalf("  %s: users file error: %v", name, err)
	}

	naliases, err := s.AddAliasesFile(name, dir+"/aliases")
	if err != nil {
		// If there's an error loading aliases, fail hard to make sure this is
		// noticed and fixed as soon as it happens.
		log.Fatalf("  %s: aliases file error: %v", name, err)
	}

	ndkim, err := loadDKIM(name, dir, s)
	if err != nil {
		// DKIM errors are fatal because if the user set DKIM up, then we
		// don't want it to be failing silently, as that could cause
		// deliverability issues.
		log.Fatalf("  %s: DKIM loading error: %v", name, err)
	}

	log.Infof("  %s (%d users, %d aliases, %d DKIM keys)",
		name, nusers, naliases, ndkim)
}

func loadDovecot(s *smtpsrv.Server, userdb, client string) {
	a := dovecot.NewAuth(userdb, client)
	s.SetAuthFallback(a)
	log.Infof("Fallback authenticator: %v", a)

	if err := a.Check(); err != nil {
		log.Errorf("Warning: Dovecot auth is not responding: %v", err)
	}
}

func loadDKIM(domain, dir string, s *smtpsrv.Server) (int, error) {
	glob := path.Clean(dir + "/dkim:*.pem")
	pems, err := filepath.Glob(glob)
	if err != nil {
		return 0, err
	}

	for _, pem := range pems {
		base := filepath.Base(pem)
		selector := strings.TrimPrefix(base, "dkim:")
		selector = strings.TrimSuffix(selector, ".pem")

		err = s.AddDKIMSigner(domain, selector, pem)
		if err != nil {
			return 0, err
		}
	}
	return len(pems), nil
}

// Read a directory, which must have at least some entries.
func mustReadDir(path string) []os.DirEntry {
	dirs, err := os.ReadDir(path)
	if err != nil {
		log.Fatalf("Error reading %q directory: %v", path, err)
	}
	if len(dirs) == 0 {
		log.Fatalf("No entries found in %q", path)
	}

	return dirs
}
